File: DFS-BFS/DFSandBFS.js
function solution(m, n, v, connected) {
  const map = Array.from(Array(1001), () => Array(1001).fill(0));
  const visited = Array.from({ length: map.length }, () => false);
  let answerOfDfs = [];
  let answerOfBfs = [];

  // mapping
  connected.forEach((value) => {
    let x = value[0];
    let y = value[1];

    map[x][y] = 1;
    map[y][x] = 1;
  });

  function bfs(start) {
    function isEmpty(arr) {
      if (arr.length === 0) return true;
      return false;
    }

    let q = [];
    q.push(start);
    answerOfBfs.push(start);
    visited[start] = true;

    while (!isEmpty(q)) {
      let headNode = q.shift();
      for (let i = 1; i < map[headNode].length; i++) {
        if (map[headNode][i] && !visited[i]) {
          q.push(i);
          answerOfBfs.push(i);
          visited[i] = true;
        }
      }
    }
    return answerOfBfs;
  }

  function dfs(v, n) {
    answerOfDfs.push(v);
    visited[v] = true;

    for (let i = 1; i < map[v].length; i++) {
      if (map[v][i] && !visited[i]) {
        return dfs(i, n + 1);
      }
    }

    return answerOfDfs;
  }

  const resultOfBfs = bfs(v);
  const resultOfDfs = dfs(v, 0);
}

/*
let visited = Arry.from({ length: map.length }, () => false);

class Queue {
  constructor() {
    this.arr = [];
  }

  pushTail(item) {
    this.arr.push(item);
    console.log(item);
  }

  popHead() {
    return this.arr.shift();
  }

  isEmpty() {
    // queue가 비었는지 확인
    if (this.arr.length === 0) return true;
    return false;
  }
}

function bfs(n) {
  let queue = new Queue();
  queue.pushTail(n);
  visited[n] = true;

  while (!queue.isEmpty()) {
    let head = queue.popHead(); // Task 1 : queue에서 head를 pop
    for (let i = 1; i < map[head].length; i++) {
      // Task 2 : head에서 갈 수 있으면서, 방문한 적 없는 node를 방문
      if (map[head][i] && !visited[i]) {
        queue.pushTail(i);
        visited[i] = true;
      }
    }
  }
}

function solution() {
  let test = bfs(1);
  console.log(test);
}
*/

File: DFS-BFS/DFSandBFS.py
import sys
N, M, V = map(int, sys.stdin.readline().split())
connected_info = [[int(i) for i in sys.stdin.readline().split()]
                  for y in range(M)]


def get_board():
    board = [[0 for col in range(N + 1)] for row in range(N + 1)]
    visited = [False for boolean in range(N + 1)]

    for info in connected_info:

        x = info[0]
        y = info[1]

        board[x][y] = 1
        board[y][x] = 1

    return board, visited


def isEmpty(arr):
    if len(arr) == 0:
        return True
    return False


board, visited = get_board()  # for DFS


def dfs(V):

    print(V, end=" ")
    visited[V] = True

    for i in range(1, len(board)):
        if board[V][i] and not visited[i]:
            return dfs(i)

    return


def bfs(V):
    board, visited = get_board()  # for BFS

    queue = []
    print(V, end=" ")
    queue.append(V)
    visited[V] = True

    while not isEmpty(queue):
        head = queue.pop(0)
        for i in range(1, len(board)):
            if board[head][i] and not visited[i]:
                print(i, end=" ")
                queue.append(i)
                visited[i] = True


dfs(V)
print()
bfs(V)

# 파이썬에서 *를 통해 2차원 배열을 형성하면 나중에 그 배열에 []통해 값의 변경을 할 경우 배열 내의 모든값이 다 수정되버림 -> 정확한 메커니즘은 모르지만 *인 것만은 확실
